rfcs/text/0015_nodejs_clustering.md

```diff
@@ -0,0 +1,526 @@
+[2021-03-02T10:23:41.903+01:00][WARN ][savedobjects-service] Skipping Saved Object migrations on startup. Note: Individual documents will still be migrated when read or written.
+```
+
+The workers logs are interleaved, and, most importantly, there is no way to see which process each log entry is coming from. 
+We will need to address that.
+
+Our options are:
+
+- Have a distinct logging configuration for each worker
+
+We could do that by automatically adding a suffix depending on the process to the appropriate file appenders configuration. 
+Note that this doesn’t solve the problem for the console appender, which is probably a no-go.
+
+- Add the process info to the log pattern and output it with the log message
+
+We could add the process name information to the log messages, and add a new conversion to be able to display it with 
+the pattern layout, such as `%process` for example.
+
+the default pattern could evolve to (ideally, only when clustering is enabled)
+
+```
+[%date][%level][%process][%logger] %message
+```
+
+The logging output would then look like
+
+```
+[2021-03-02T10:23:41.834+01:00][INFO ][worker-1][plugins-service] Plugin initialization disabled.
+[2021-03-02T10:23:41.840+01:00][INFO ][worker-2][plugins-service] Plugin initialization disabled.
+```
+
+This will require some non-trivial changes in the logging system implementation, as currently the BaseLogger has no logic to 
+enhance the log record with context information before forwarding it to its appenders, but this still
+seems to be the best solution. (This may even be the first step to implement our MDC)
+
+Notes:
+- Even if we add the `%process` pattern, do we want to also allow users to specify per-worker log files? 
+
+### The rolling-file appender
+
+The rolling process of the `rolling-file` appender is going to be problematic in clustered mode, as it will cause 
+concurrency issues during the rolling. We need to find a way to have this rolling stage clustered-proof.
+
+Identified options are:
+
+- have the rolling file appenders coordinate themselves when rolling
+
+By using a broadcast message based mutex mechanism, the appenders could acquire a ‘lock’ to roll a specific file, and
+notify other workers when the rolling is complete (quite similar to what we want to do with SO migration for example).
+
+- have the coordinator process perform the rolling
+
+Another option would be to have the coordinator perform the rotation instead. When a rolling is required, the appender 
+would send a message to the coordinator, which would perform the rolling and notify the workers once the operation is complete. 
+
+Note that this option is even more complicated than the previous one, as it forces to move the rolling implementation 
+outside of the appender, without any significant upsides identified.
+
+- centralize the logging system in the coordinator (and have the workers send messages to the coordinator when using the logging system)
```

Comment:
Would we have to centralize it in the coordinator process? Wouldn't it be simpler to just forward the logs to the `main` Kibana worker? That way we wouldn't have to two very different implementations or pull any logging code outside of Core, but can still benefit from having a single process doing the writing.

Reply:
> That way we wouldn't have to two very different implementations or pull any logging code outside of Core

I don't think we need to pull logging code outside of core in either option? The coordinator class belongs to Core too.

Regarding the implementations, both approaches have the same complexity I believe, as both requires to short circuit the current / default logging system implementation to send messages instead of invoking the logging logic when in clustering mode.

Also, the coordinator will probably need to also output logs, and it would be awkward to have the coordinator rely on the main worker to process its own logs I think?